### ISSUE TITLE
mocking functions from a shared library

### ISSUE THREAD

**Issue by kerby82** [jenkinsci/JenkinsPipelineUnit#357]
I'm testing groovy file that uses a function from a shared library

foo.groovy
```
def call(branch) {
if (!git.fetch(branch)) {
error "Error fetching ${branch}"
}
}
```

I want to test it in isolation mode without the need of registering the SharedLibrary.

But I got the following error:
`groovy.lang.MissingPropertyException: No such property: git for class: foo`

The test just does this:
```
class FooTest extends BaseTest {


  def foo
  

  @Before
  void setUp() throws Exception {
    super.setUp()
    foo = loadScript("foo.groovy")
    helper.registerAllowedMethod('fetch', [String], {true})
  }

  @Test
  void fooTest() throws Exception {
    foo("master")
    printCallStack()
  }

**Comment by nre-ableton** [comment]
Is `foo.groovy` a singleton, for example, located in `vars/foo.groovy` in your library?

**Comment by kerby82** [comment]
Yes is located in the `vars` but the git.fetch is located in a shared-library that is imported by the Jenkinsfile that is using foo. And my goal is to test foo in isolation mode.

**Comment by nre-ableton** [comment]
I think you need to mock the library import, in that case.

**Comment by kerby82** [comment]
What do you mean?
In this case my problem is I'm not able to mockup `git.fetch` because it does not find the property git, but I don't know how to set such a property so I can mock the fetch method.

**Comment by microhod** [comment]
I think because you're wanting to mock all of `git` you need to use `binding.setVariable` instead of `helper.registerAllowedMethod`. so you'd have:
```groovy
  @Before
  void setUp() throws Exception {
    super.setUp()
    foo = loadScript("foo.groovy")
    binding.setVariable('git', new MockGit())
  }
```
Where `MockGit` can be just a mock class you can create in your test to mock out git.

Or you could probably just set it to a map with a closure if you don't want to bother with a whole mock class:
```groovy
binding.setVariable('git', ['fetch': {s -> true}])
```

**Comment by waytoharish** [comment]
i have a Utility.groovy which is inside the `src.util.data` package and when using in  the `var/MyJenkins.groovy` . I am trying to mock the same but not able to mock the methods inside the Utility.groovy. 
Please could someone help here.
